src/security/Handshake.cc

```diff
@@ -285,12 +286,18 @@ Security::HandshakeParser::parseChangeCipherCpecMessage()
 {
     Must(currentContentType == ContentType::ctChangeCipherSpec);
     // We are currently ignoring Change Cipher Spec Protocol messages.
-    skipMessage("ChangeCipherCpec msg [fragment]");
+    skipMessage("ChangeCipherSpec msg [fragment]");
 
-    // Everything after the ChangeCipherCpec message may be encrypted.
+    // In TLS v1.3, a dummy ChangeCipherSpec implies Middlebox Compatibility
+    // Mode. In earlier TLS versions, ChangeCipherSpec starts (encrypted)
+    // session resumption
+    const bool tlsv13OrLater = details->tlsSupportedVersion.protocol == AnyP::PROTO_TLS && details->tlsSupportedVersion >= AnyP::ProtocolVersion(AnyP::PROTO_TLS, 1, 3);
+    if (!tlsv13OrLater)
+        resumingSession = true;
+
+    // Everything after the ChangeCipherSpec message may be encrypted.
     // Continuing parsing is pointless. Stop here.
```

Comment:
The following is specific to TLS v1.3 (and later):

Is further parsing pointless here? Why? Because the subsequent messages must be encrypted, just like in earlier protocol versions? What RFC section implies that? 

I am asking these questions, in part, because RFC [8446](https://tools.ietf.org/html/rfc8446#appendix-D.4) says `Either side can send change_cipher_spec at any time during the handshake, as they must be ignored by the peer` and we are not quite ignoring it -- we _stop_ parsing. How do we justify that?

Reply:
The first ChangeCipherSpec is comming after we parsed server hello which included in first HandshakeMessages. The following messages may have any number of ChangeCipherSpec, but we are not interested any more. We can not retrieve more information from TLSv1.3 server hello messages.

It is making sense to ignore ChangeCipherSpec messages only if the ServerHello message is split across more than one Handshake messages, which separated by ChangeCipherSpec messages. But this is not possible.

Reply:
> RFC: Either side can send change_cipher_spec at any time during the handshake
> Christos: The first ChangeCipherSpec is comming after we parsed server hello

Why not _before_ the server Hello, like the RFC allows? I am sure that "after" is what you see in your tests, but if the RFC says "anytime" and "ignore", should not the code be more careful in its assumptions and follow the RFC more closely?

> We can not retrieve more information from TLSv1.3 server hello messages.

Why not? Because the code (assumes that we have) parsed the TLS server Hello already? Then why are we not stopping right after we parsed that Hello? Something is missing either in the code itself or in our logic that tries to justify that code...

Reply:
> Why not before the server Hello, like the RFC allows? I am sure that "after" is what you see in your tests, but if the RFC says "anytime" and "ignore", should not the code be more careful in its assumptions and follow the RFC more closely?

The true is that there is not any requirement that the ServerHello must be the first message. The RFC describes also the case of the HelloRetryRequest message which may sent by server and which is not handled by squid.

However initial idea of this code is to take advantage of the compatibility mode which is used by existing TLS client implementations.

The RFC in appendix D.4 says:
 > The server sends a dummy change_cipher_spec record immediately
      after its first handshake message.  This may either be after a
      ServerHello or a HelloRetryRequest.

  > When put together, these changes make the TLS 1.3 handshake resemble
   TLS 1.2 session resumption, which improves the chance of successfully
   connecting through middleboxes.  This "compatibility mode" is
   partially negotiated: the client can opt to provide a session ID or
   not, and the server has to echo it.  Either side can send
   change_cipher_spec at any time during the handshake, as they must be
   ignored by the peer, but if the client sends a non-empty session ID,
   the server MUST send the change_cipher_spec as described in this
   appendix.

My sense is that the server may sent a change_cipher_spec anywhere only if the compatibility mode is not used, when the client sends an empty sesion ID.  

So to handle correctly cases where an empty change_cipher_spec is send anywhere, we must also check the client sessionID ad client tls version. We need to pass at least this info (client is tlsv1.3 and sessionID is empty or not) to HandshakeParser as parameter.

 > Why not? Because the code (assumes that we have) parsed the TLS server Hello already? Then why are we not stopping right after we parsed that Hello

For TLSv1.3, we can stop after parsed the server hello. Because any data we are going to receive after server hello  are encrypted. Probably in bump/stare modes we are interested to get more (encrypted) data to feed openSSL subsystem (which can extract certificates and other info) but I do not know how we can ensure the end of these data.
For tlsv1.2 the end of the interested data is the CCS. 

But for now I think, if we detect a TLSv1.3 server hello, we can just stop after parsing server hello. It is easy to change that, just a check for TLSv1.3 protocol inside parseServerHelloHandshakeMessage which sets the HandshakeParser::done member.

Reply:
> Why not (ChangeCipherSpec message) before the server Hello, like the RFC allows?

I did not make any change to fix this.  I need your confirmation before proceed to this fix. My comments about:
  - For TLSv1.2 or earlier we are checking for CCS to detect resumed sessions. 
 - For TLSv1.2 or earlier releases the CCS before server hello is not allowed. 
  - A server may send a CCS before hello only if the client is sent a TLSv1.3 hello message. We should check for this too. 
  - I do not believe that there are many servers which may sent CCS before server hello. Only if it is a try to break a proxy policy. Squid will stop parsing message without extracting any info from server hello. The only info can extracted from squid is TLS versions and other TLS features. The server certificate is not available in TLSv1.3. I think not a critical info is missing here. 

Also check my previous comment about.

> Then why are we not stopping right after we parsed that Hello?

Commit 2693b6fd2 fixes this for TLSv1.3 messages.

Also please check my commit c39c38f67 which fixes my previous commit c296230f

Reply:
There are many nuances here, but most of them are either not important or out of scope. We should focus on this PR problem/scope. We just need to be careful not to say/imply something that is false or inaccurate. I suggest the following simple solution:

1. For TLS v1.2 and earlier, set `resumingSession` because that is what `ChangeCipherSpec` means for those TLS versions. We should also set `done` in the _same_`if` statement -- we are done because we have determined that the session is being resumed.
2. For TLS v1.3 and later, set `done` at the end of parsing Hello because we do not expect any more (parseable) data. We should ignore ChangeCipherSpec for these protocol versions (at least until we need to detect Compatibility Mode for some reason).

If the above plan makes sense to you, please adjust the code accordingly (AFAICT, the current code is already pretty close to the above sketch!). Otherwise, please let me know what I missed.

Reply:
please check my commit 7320215a3a

This is fixes ChangeCipherSpec messages parsing to check if we know that we are parsing a TLSv1.2 or earlier TLS hello handshake messages. If yes mark resumming session and stop parsing. The effect of the above is that if a ChangeCipherSpec comes before the  hello handshake message, the handshake parser ignore it, even if we do
not know (yet) that we are parsing a TLSv1.3 or not.

Reply:
Commit 7320215 looks like a good step forward. We need to add a comment to explain why we are ignoring CSS for TLS v1.3 and another comment to explain why we are ignoring CSS when we have not received Hello yet (and, hence, do not know even the TLS version yet). AFAICT, those comments will say something like this:

```C++
// In TLS v1.2 and earlier, ChangeCipherSpec is sent after Hello (so the
// version is known by now) and indicates session resumption.
// In later TLS versions, ChangeCipherSpec may be sent before and after Hello,
// but it is no longer used for session resumption and should be ignored.
```

Any other changes that you would like to make to this PR?

Reply:
I comited 2 fixes. I hope they are OK.
If you think they are OK, I think this branch is ready for an official review.